Worker.requestAnimationFrame spec

fserb opened the pull request:
Draft change to: https://github.com/whatwg/html/issues/3587


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/3677/canvas.html" title="Last updated on Aug 24, 2018, 10:46 AM GMT (5a54407)">/canvas.html</a>  ( <a href="https://whatpr.org/html/3677/b71f885...5a54407/canvas.html" title="Last updated on Aug 24, 2018, 10:46 AM GMT (5a54407)">diff</a> )
<a href="https://whatpr.org/html/3677/embedded-content.html" title="Last updated on Aug 24, 2018, 10:46 AM GMT (5a54407)">/embedded-content.html</a>  ( <a href="https://whatpr.org/html/3677/b71f885...5a54407/embedded-content.html" title="Last updated on Aug 24, 2018, 10:46 AM GMT (5a54407)">diff</a> )
<a href="https://whatpr.org/html/3677/imagebitmap-and-animations.html" title="Last updated on Aug 24, 2018, 10:46 AM GMT (5a54407)">/imagebitmap-and-animations.html</a>  ( <a href="https://whatpr.org/html/3677/b71f885...5a54407/imagebitmap-and-animations.html" title="Last updated on Aug 24, 2018, 10:46 AM GMT (5a54407)">diff</a> )
<a href="https://whatpr.org/html/3677/infrastructure.html" title="Last updated on Aug 24, 2018, 10:46 AM GMT (5a54407)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/3677/b71f885...5a54407/infrastructure.html" title="Last updated on Aug 24, 2018, 10:46 AM GMT (5a54407)">diff</a> )
<a href="https://whatpr.org/html/3677/timers-and-user-prompts.html" title="Last updated on Aug 24, 2018, 10:46 AM GMT (5a54407)">/timers-and-user-prompts.html</a>  ( <a href="https://whatpr.org/html/3677/b71f885...5a54407/timers-and-user-prompts.html" title="Last updated on Aug 24, 2018, 10:46 AM GMT (5a54407)">diff</a> )
<a href="https://whatpr.org/html/3677/webappapis.html" title="Last updated on Aug 24, 2018, 10:46 AM GMT (5a54407)">/webappapis.html</a>  ( <a href="https://whatpr.org/html/3677/b71f885...5a54407/webappapis.html" title="Last updated on Aug 24, 2018, 10:46 AM GMT (5a54407)">diff</a> )
<a href="https://whatpr.org/html/3677/window-object.html" title="Last updated on Aug 24, 2018, 10:46 AM GMT (5a54407)">/window-object.html</a>  ( <a href="https://whatpr.org/html/3677/b71f885...5a54407/window-object.html" title="Last updated on Aug 24, 2018, 10:46 AM GMT (5a54407)">diff</a> )

fserb commented:
cc @whatwg/canvas

junov commented on source:
```diff
@@ -93344,44 +93358,52 @@ loadMySprites().then(runDemo);</pre>
 
-   <li><p>Append the method's argument to <var>document</var>'s <span>list of animation frame
-   callbacks</span>, associated with <var>document</var>'s <span>animation frame callback
+   <li><p>Append the method's argument to <var>object</var>'s <span>list of animation frame
```
In the case of the Window interface, these internal slots are being moved from the document object to the window object.  I think this is fine, but it raises the question: Why were they on the document in the first place?  Is there a special use case where this makes a difference?

annevk replied:
Yeah, initial about:blank replaces the document but keeps the window around. Is that tested? (There's also `document.open()` which keeps the document, but allocates a new window, but that is not consistent across browsers and will likely go away.)

fserb replied:
That makes sense. I've introduced a new description for *target object* to clear this up.

junov commented:
Other that the question raised aboved, this looks good to me.

annevk commented on source:
```diff
@@ -65418,7 +65418,8 @@ interface <dfn>OffscreenCanvasRenderingContext2D</dfn> {
     the bitmap of the <span data-x="offscreencanvas-placeholder">placeholder <code>canvas</code>
     element</span> of the <span>associated <code>OffscreenCanvas</code> object</span>. The copy
-    operation is asynchronous.</p>
+    operation is synchronous. Calling this method is <!--non-normative-->optional for the
```
How about "not needed" instead of "optional"?

fserb replied:
done

annevk commented on source:
```diff
@@ -65473,7 +65474,7 @@ interface <dfn>OffscreenCanvasRenderingContext2D</dfn> {
 
   <p>The <dfn><code data-x="offscreencontext2d-commit">commit()</code></dfn> method, when invoked,
-  must run the following steps:</p>
+  has to run the following steps:</p>
```
Why did you change this?

fserb replied:
lint complained about it, maybe I misunderstood the line. Reverted.

annevk commented on source:
```diff
@@ -89102,13 +89103,26 @@ dictionary <dfn>PromiseRejectionEventInit</dfn> : <span>EventInit</span> {
 
-   <li><p>If this is a <a href="#workers">worker</a> <span>event loop</span> (i.e. one running for a
-   <code>WorkerGlobalScope</code>), but there are no <span data-x="concept-task">tasks</span> in the
+   <li><p>If that is a <a href="#workers">worker</a> <span>event loop</span> (i.e. one running for a
```
The formatting in this block is a little off.

fserb replied:
done.

annevk commented on source:
```diff
@@ -89102,13 +89103,26 @@ dictionary <dfn>PromiseRejectionEventInit</dfn> : <span>EventInit</span> {
+         passing in <var>now</var> as the timestamp.</p></li>
+
+      <li><p>Update the rendering or user interface of that <code>Worker</code> to reflect the
+        current state.</p></li>
```
Workers don't have a user interface so I don't think we should mention that here.

fserb replied:
done

annevk commented on source:
```diff
@@ -121656,7 +121678,7 @@ INSERT INTERFACES HERE
   <div itemscope itemtype="http://n.whatwg.org/work">
    <p>The <code data-x="">&lt;flag-icon></code> example was inspired by <a itemprop="work"
-   href="https://github.com/stevenrskelton/flag-icon">a custom element</a> by <a
+   href="https://github.com/stevenrskelton/flag-icon">a custom Element</a> by <a
```
This was probably not intended?

annevk commented on source:
```diff
@@ -93392,18 +93414,18 @@ loadMySprites().then(runDemo);</pre>
 
   </ol>
 
+  <p>Currently <code>Window</code> and <code>WorkerGlobalScope</code> implement <code>AnimationFrameProvider</code>.</p>
```
Are there plans for other interfaces? Note that they include it, not implement.

annevk commented on source:
```diff
@@ -78185,8 +78186,8 @@ interface <dfn data-export="" data-dfn-type="interface">Window</dfn> : <span>Eve
-  unsigned long <span data-x="dom-window-requestAnimationFrame">requestAnimationFrame</span>(<span>FrameRequestCallback</span> callback);
-  void <span data-x="dom-window-cancelAnimationFrame">cancelAnimationFrame</span>(unsigned long handle);
+  unsigned long <span data-x="AnimationFrameProvider-requestAnimationFrame">requestAnimationFrame</span>(<span>FrameRequestCallback</span> callback);
+  void <span data-x="AnimationFrameProvider-cancelAnimationFrame">cancelAnimationFrame</span>(unsigned long handle);
```
Why do you keep these around as well as defining a mixin with a different signature? That contradicts the statement below that Window and WorkerGlobalScope include the same mixin.

annevk commented on source:
```diff
@@ -93392,18 +93414,18 @@ loadMySprites().then(runDemo);</pre>
+   <li><p>Set <var>object</var>'s <span>list of animation frame callbacks</span> to the empty list.</p></li>
 
    <li><p>For each entry in <var>callbacks</var>, in order: <span data-x="es-invoking-callback-functions">invoke the callback</span>, passing
    <var>now</var> as the only argument, and if an exception is thrown, <span>report the
```
You define the callback as taking two arguments. This contradicts that.

fserb replied:
ahn, sorry. The IDL update was wrong. In the future there may be a second argument, but not now.

fserb commented:
Done. PTAL. :)

annevk commented on source:
```diff
@@ -93344,44 +93354,55 @@ loadMySprites().then(runDemo);</pre>
+interface AnimationFrameProvider {
+  unsigned long requestAnimationFrame(FrameRequestCallback callback);
+  void cancelAnimationFrame(unsigned long handle);
+}
```
You need to add something like this here:
```
Window includes AnimationFrameProvider;
WorkerGlobalScope includes AnimationFrameProvider;
```

annevk commented on source:
```diff
@@ -93391,19 +93412,20 @@ loadMySprites().then(runDemo);</pre>
 
   </ol>
 
+  <p>Currently <code>Window</code> and <code>WorkerGlobalScope</code> include the <code>AnimationFrameProvider</code> interface.</p>
```
Is there a plan beyond these objects? If not, should we just put them on WindowOrWorkerGlobalScope?

annevk replied:
You answered this above already with WebXR. I guess that means we best keep it separate from Window/WorkerGlobalScope.

fserb replied:
k

annevk commented on source:
```diff
@@ -89102,13 +89100,25 @@ dictionary <dfn>PromiseRejectionEventInit</dfn> : <span>EventInit</span> {
+    <li><p>Update the rendering of that <code>Worker</code> to reflect the current state.</p></li>
+
+    <li><p class="note">The same note described above regarding update frequency applies
+    to <code>Worker</code>.</p></li>
```
That note is about "would not benefit from having their rendering updated" which isn't used here. I guess this should actually be more complex so we don't run this step for each task necessarily?

fserb replied:
I'm not sure what you mean by more complex. I've updated the note slightly. Would that work?

annevk commented on source:
```diff
@@ -93344,44 +93354,55 @@ loadMySprites().then(runDemo);</pre>
+}
+  </pre>
+
+  <p>Each <var>object</var> implementing the <code>AnimationFrameProvider</code> interface
```
I don't think this should be a variable since you never really assigned it. In the algorithm below you can use "this object" instead. Here you can use "object".

annevk replied:
I guess you also need to change the language here since these days objects include mixins.

annevk commented on source:
```diff
@@ -93344,44 +93354,55 @@ loadMySprites().then(runDemo);</pre>
+
+callback FrameRequestCallback = void (DOMHighResTimeStamp);
+
+interface AnimationFrameProvider {
```
interface mixin

fserb replied:
done

annevk commented on source:
```diff
@@ -93344,44 +93354,55 @@ loadMySprites().then(runDemo);</pre>
   <h3>Animation frames</h3>
 
-  <p>Each <code>Document</code> has a <dfn>list of animation frame callbacks</dfn>, which must be
+  <p>Some objects implement the <dfn>AnimationFrameProvider</dfn> mixin.</p>
```
interface mixin

fserb replied:
done

annevk commented on source:
```diff
@@ -93344,44 +93354,55 @@ loadMySprites().then(runDemo);</pre>
+  <pre class="idl">
+ interface FrameRequestData {
+  // Empty base interface
+}
```
This appears unused. Remove it?

annevk commented on source:
```diff
@@ -93344,44 +93354,55 @@ loadMySprites().then(runDemo);</pre>
 
-   <li><p>Let <var>document</var> be this <code>Window</code> object's <span
-   data-x="concept-document-window">associated <code>Document</code></span>.</p></li>
+   <li><p>Consider the <var>target object</var> the <var>object</var> that stores <code>AnimationFrameProvider</code>'s internal state. If the interface is implemented by <code>Window</code>, then <var>target object</var> is <code>Document</code>. If is implemented by <code>WorkerGlobalScope</code>, then <var>target object</var> is <code>WorkerGlobalScope</code>.</p></li>
```
You need to keep the language of "associated Document". Otherwise it's unclear where the document comes from.

annevk replied:
And also not use "implemented", but "includes".

fserb replied:
done

annevk commented:
Yeah, so the worker event loop changes seems incorrect. In particular, you run the substeps for each task processed by a worker, whereas you only want to run them sometimes, say, at 60Hz.

In particular, you want something like substep 3 of "Update the rendering" (the one for documents) where you do an early return if the worker does not require a rendering update. And there you want to attach the note.

Is this logic of when you skip a rendering update for a worker tied to its owners by the way? If it has an owner document that is not displayed, is it valid to not give it rendering updates?

Should this be exposed in service workers by the way? This API does not make much sense to me there given how it be quit early and such.

fserb commented:
Gotcha! :)
I cleared up all references to WorkerGlobalScope to be DedicatedWorkerGlobalScope. (which clears the ServiceWorker thing).
And I think I understood what you meant by the event loop thing and fixed it.

annevk commented on source:
```diff
@@ -93273,19 +93302,20 @@ loadMySprites().then(runDemo);</pre>
 
   </ol>
 
+  <p>Currently <code>Window</code> and <code>DedicatedWorkerGlobalScope</code> include the <code>AnimationFrameProvider</code> interface.</p>
```
I suggest we drop this sentence since it's already clear from the IDL.

fserb replied:
Done

annevk commented on source:
```diff
@@ -88985,12 +88983,30 @@ dictionary <dfn>PromiseRejectionEventInit</dfn> : <span>EventInit</span> {
+
+       <li><p>Update the rendering of that <code>Worker</code> to reflect the current state.</p></li>
+      </ol>
+     </li>
```
These first two steps should be conditional on it being a dedicated worker.

fserb replied:
Done.

annevk commented on source:
```diff
@@ -88985,12 +88983,30 @@ dictionary <dfn>PromiseRejectionEventInit</dfn> : <span>EventInit</span> {
+     <li><p>If the user agent believes that the <code>Worker</code> would benefit from having their
+      rendering updated at this time, then:
+      <ol>
+       <li><p><span>run the animation frame callbacks</span> for that <code>Worker</code>,
```
This should be `DedicatedWorkerGlobalScope` I think.

fserb replied:
done

fserb commented:
All done. :)

annevk commented:
Where is the `commit()` aspect of the change proposal? As far as I can tell all it does thus far is making `requestAnimationFrame()` available in a dedicated worker.

How does it relate btw that `OffscreenCanvas` is available in all workers? Isn't that problematic?

fserb commented:
Yes, I'll send a `commit()` one soon.

For the `transferControlToOffscreen()`, yes. Nothing prevents it yet on the spec, as far as I understand.
I'd claim that without a `DedicatedWorker`, the only way a `OffscreenCanvas` can be updated is with a direct `commit()` (the new blocking one). This RAF + rendering event loop model just allows `DedicatedWorker` to make `OffscreenCanvas` behave nicely.

fserb commented:
I'll give a try at the formatting guidelines, although I thought the linter got them all. I can easily update to #3707 once it lands.

fserb commented:
Fixed formatting.

fserb commented:
`commit()` changes were already changed when we removed the `Promise` work.

annevk commented on source:
```diff
@@ -88980,13 +88978,33 @@ dictionary <dfn>PromiseRejectionEventInit</dfn> : <span>EventInit</span> {
+       <code>DedicatedWorkerGlobalScope</code>, passing in <var>now</var> as the
+       timestamp.</p></li>
+
+       <li><p>Update the rendering of that <code>Worker</code> to reflect the current state.</p>
```
This should talk about the dedicated worker, not the object used to talk to a dedicated worker.

fserb replied:
done

annevk commented on source:
```diff
@@ -88980,13 +88978,33 @@ dictionary <dfn>PromiseRejectionEventInit</dfn> : <span>EventInit</span> {
+     described in the <a href="#workers">Web workers</a> section below.</p></li>
+    </ol>
+    <p class="note">Similar to the <span data-x="nested browsing context">nested browsing contexts</span>
+    note above, a user agent can consider updating <code>Worker</code> rendering infrequently or never.</p>
```
Same.

fserb replied:
done

annevk commented on source:
```diff
@@ -93269,16 +93303,19 @@ loadMySprites().then(runDemo);</pre>
   steps:</p>
 
   <ol>
+   <li><p>Consider <var>target object</var> as above.</p></li>
```
This is too informal. We should either duplicate or create some kind of abstraction.

fserb replied:
I've moved the notion of a target object above.

annevk commented on source:
```diff
@@ -88980,13 +88978,33 @@ dictionary <dfn>PromiseRejectionEventInit</dfn> : <span>EventInit</span> {
+   	a <code>WorkerGlobalScope</code>):
+    <ol>
+     <li><p>Let <var>now</var> be the value that would be returned by the <code>Performance</code>
+     object's <code data-x="dom-Performance-now">now()</code> method. <ref spec=HRT></p>
```
This should use the primitive from the other PR as discussed. This step should also be conditional on it being a dedicated worker.

fserb replied:
I'll update once that one is submitted.

fserb commented:
@mstange @nical ?

fserb commented:
@mephisto41 ?

chrishtr commented:
@jdashg could you give feedback? Thanks.

jdashg commented:
LGTM

cvan commented on source:
```diff
@@ -88980,13 +88978,35 @@ dictionary <dfn>PromiseRejectionEventInit</dfn> : <span>EventInit</span> {
+   	a <code>WorkerGlobalScope</code>):
+    <ol>
+     <li><p>If this is a <code>DedicatedWorkerGlobalScope</code> and the user agent believes that
+      it would benefit from having their rendering updated at this time, then:
```
nit: can you close the `</p>` here?

cvan replied:
nit: `their` -> `its`

cvan replied:
`,then` -> `, then it must`?

fserb replied:
done

cvan commented on source:
```diff
@@ -65380,7 +65380,8 @@ interface <dfn>OffscreenCanvasRenderingContext2D</dfn> {
     element</span> of the <span>associated <code>OffscreenCanvas</code> object</span>. The copy
-    operation is asynchronous.</p>
+    operation is synchronous. Calling this method is not needed for the transfer, since it happens
+    automatically during the event loop execution.</p>
```
nit: `event loop` -> `<span>event loop</span>` (per elsewhere)

fserb replied:
done

cvan commented on source:
```diff
@@ -88980,13 +88978,35 @@ dictionary <dfn>PromiseRejectionEventInit</dfn> : <span>EventInit</span> {
-   loop</span>, aborting these steps, resuming the <span>run a worker</span> steps described in the
-   <a href="#workers">Web workers</a> section below.</p></li>
+   <li><p>If this is a <a href="#workers">worker</a> <span>event loop</span> (i.e. one running for
+   	a <code>WorkerGlobalScope</code>):
```
nit: can you close the `</p>` here?

cvan replied:
nit: `(i.e. ` -> `(i.e., `

fserb replied:
done

cvan commented on source:
```diff
@@ -93222,45 +93242,58 @@ loadMySprites().then(runDemo);</pre>
+  <p>Each object that provides an <code>AnimationFrameProvider</code> also has a
+  <var>target object</var> that stores the animation frame provider's internal state. If the
+  interface mixin is included by <code>Window</code>, then <var>target object</var> is this
+  <code>Window</code> <span data-x="concept-document-window">associated <code>Document</code>
```
`<code>Window</code>` -> `<code>Window</code>'s`

fserb replied:
done

cvan commented on source:
```diff
@@ -93222,45 +93242,58 @@ loadMySprites().then(runDemo);</pre>
+  </pre>
 
-   <li><p>Increment <var>document</var>'s <span>animation frame callback identifier</span> by
+  <p>Each object including the <code>AnimationFrameProvider</code> interface has a <dfn>list of
```
nit: `Each object including the <code>AnimationFrameProvider</code> interface has` -> `Each object, including the <code>AnimationFrameProvider</code> interface, has`

fserb replied:
That's not exactly what we meant. But made the language better.

cvan commented on source:
```diff
@@ -93222,45 +93242,58 @@ loadMySprites().then(runDemo);</pre>
-  data-x="dom-window-requestAnimationFrame">requestAnimationFrame()</code></dfn> method is called,
-  the user agent must run the following steps:</p>
+  <pre class="idl">
+callback FrameRequestCallback = void (DOMHighResTimeStamp);
```
should `DOMHighResTimeStamp` be `DOMHighResTimeStamp time`?

fserb replied:
yep. Done.

cvan commented on source:
```diff
@@ -93269,16 +93302,17 @@ loadMySprites().then(runDemo);</pre>
-   <li><p>Set <var>doc</var>'s <span>list of animation frame callbacks</span> to the empty
+   <li><p>Let <var>callbacks</var> be a list of the entries in <var>target object</var>'s
+   <span>list of animation frame callbacks</span>, in the order in which they were added to the
    list.</p></li>
```
`list` -> `<var>callbacks</var>`?

fserb replied:
done

cvan commented on source:
```diff
@@ -93269,16 +93302,17 @@ loadMySprites().then(runDemo);</pre>
-   <var>now</var> as the only argument, and if an exception is thrown, <span>report the
-   exception</span>. <ref spec=WEBIDL></p></li>
+   <li><p>Set <var>target object</var>'s <span>list of animation frame callbacks</span> to the
+   empty list.</p></li>
```
`the empty list`? or `an empty list`?

fserb replied:
done

cvan commented on source:
```diff
@@ -93269,16 +93302,17 @@ loadMySprites().then(runDemo);</pre>
+
+   <li><p>For each entry in <var>callbacks</var>, in order:
+   <span data-x="es-invoking-callback-functions">invoke the callback</span>, passing <var>now</var>
+   as the only argument, and if an exception is thrown, <span>report the exception</span>.
```
if an exception is thrown, what happens to the overall `target object`?

fserb replied:
well, any exception can be thrown from the Javascript code.
I'm not sure what I'm supposed to mention here. Those are not UA-side exceptions, it's just meant to say that it's valid for the js code to throw exceptions during the callback. It's the same language that we had before.

cvan commented on source:
```diff
@@ -93269,16 +93302,17 @@ loadMySprites().then(runDemo);</pre>
-   animation frame callbacks</span>, in the order in which they were added to the list.</p></li>
-
-   <li><p>Set <var>doc</var>'s <span>list of animation frame callbacks</span> to the empty
+   <li><p>Let <var>callbacks</var> be a list of the entries in <var>target object</var>'s
```
would it make sense to add a separate, initial step for `Let <var>callbacks</var> be an empty list.`? and then create a separate step with an ordered list for the substeps, à la `To append an entry to <var>callbacks</var> list, run these steps:`?

cvan replied:
disregard this. I didn't see this is [defined above for `requestAnimationFrame`](https://github.com/whatwg/html/pull/3677/files#diff-36cd38f49b9afa08222c0dc9ebfe35ebR88985)

cvan commented on source:
```diff
@@ -93269,16 +93302,17 @@ loadMySprites().then(runDemo);</pre>
   steps:</p>
```
I believe this is a typo you can remove

cvan replied:
I very likely could be reading this wrongly, but is this missing a section / paragraph?

fserb replied:
I'm not sure what you are referring to.

cvan commented on source:
```diff
@@ -93222,45 +93242,58 @@ loadMySprites().then(runDemo);</pre>
-   frame callbacks</span>.</p></li>
+   <li><p>If there is such an entry, remove it from <var>target object</var>'s <span>list of
+   animation frame callbacks</span>.</p></li>
 
```
should there be a note here about handling exceptions?

fserb replied:
not sure I get it. on `cancelAnimationFrame` there's no exception that should happen.

cvan commented on source:
```diff
@@ -93269,16 +93302,17 @@ loadMySprites().then(runDemo);</pre>
+   <li><p>Set <var>target object</var>'s <span>list of animation frame callbacks</span> to the
+   empty list.</p></li>
+
+   <li><p>For each entry in <var>callbacks</var>, in order:
```
I'm likely misreading this: is this instruction for `requestAnimationFrame` or `cancelAnimationFrame`?

fserb replied:
they are a different set of steps.
There are 3 entries:
- requestAnimationFrame
- cancelAnimationFrame
- running animation frame callbacks

they are 3 separate regions.

cvan commented on source:
```diff
@@ -93222,45 +93242,58 @@ loadMySprites().then(runDemo);</pre>
-  <p>Each <code>Document</code> has a <dfn>list of animation frame callbacks</dfn>, which must be
-  initially empty, and an <dfn>animation frame callback identifier</dfn>, which is a number which
-  must initially be zero.</p>
+  <p>Some objects include the <dfn>AnimationFrameProvider</dfn> interface mixin.</p>
```
:+1: nice to see this change remove the `Document` dependency. great work!

cvan commented on source:
```diff
@@ -88980,13 +88978,35 @@ dictionary <dfn>PromiseRejectionEventInit</dfn> : <span>EventInit</span> {
+       <code>Performance</code> object's <code data-x="dom-Performance-now">now()</code> method.
+       <ref spec=HRT></p></li>
+
+       <li><p><span>run the animation frame callbacks</span> for that
```
`Run the <span>list of animation frame callbacks</span>`?

cvan commented on source:
```diff
@@ -93222,45 +93242,58 @@ loadMySprites().then(runDemo);</pre>
-
-   <li><p>If there is such an entry, remove it from <var>document</var>'s <span>list of animation
-   frame callbacks</span>.</p></li>
+   <li><p>If there is such an entry, remove it from <var>target object</var>'s <span>list of
```
is it possible that the entry is null?

fserb replied:
it's possible that there's no entry. But current spec (and implementations) says it should fail silently when it can't find an entry.

cvan commented on source:
```diff
@@ -88980,13 +88978,35 @@ dictionary <dfn>PromiseRejectionEventInit</dfn> : <span>EventInit</span> {
+     <li><p>If there are no <span data-x="concept-task">tasks</span> in the
+     <span>event loop</span>'s <span data-x="task queue">task queues</span> and the
+     <code>DedicatedWorkerGlobalScope</code> object's <span
+     data-x="dom-WorkerGlobalScope-closing">closing</span> flag is true, then destroy the
```
should `dom-WorkerGlobalScope-closing` be `dom-DedicatedWorkerGlobalScope-closing`?

fserb replied:
there's no such a thing. closing is defined on the WorkerGlobalScope.

cvan commented on source:
```diff
@@ -93222,45 +93242,58 @@ loadMySprites().then(runDemo);</pre>
-   <li><p>Return <var>document</var>'s <span>animation frame callback identifier</span>'s current
-   value.</p></li>
+   <li><p>Return <var>target object</var>'s <span>animation frame callback identifier</span>'s
+   current value.</p></li>
```
do you want to give `current value` a `<var>`?

fserb replied:
removed "current value"

cvan commented on source:
```diff
@@ -93269,16 +93302,17 @@ loadMySprites().then(runDemo);</pre>
+   <li><p>Set <var>target object</var>'s <span>list of animation frame callbacks</span> to the
+   empty list.</p></li>
+
+   <li><p>For each entry in <var>callbacks</var>, in order:
```
forgive me in advance for the possibly unnecessary verbosity. what are your thoughts on something like this?

```
   <p><dfn>The callback processing algorithm</dfn> for a callback <var>callback</var> is as follows:</p>
   <ol>
    <li><p><span>Prepare to run a callback</span>, <var>callback</var>.</p></li>
    <li><p><span data-x="stack push">Push</span> <var>script execution context</var> onto the <span>JavaScript execution context stack</span>.</p></li>
    <li><p>Let <var>result</var> be the result of performing the <span>animation frame callback</span>.</p></li>
    <li><p>If <var>result</var> is null, then return.</p></li>
    <li><p><span data-x="es-invoking-callback-functions">Invoke</span> the callback <var>callback</var> with one argument, the value of which is a <code>DOMHighResTimeStamp</code> timestamp, <var>now</var>.</p></li>
    <li><p><span data-x="stack pop">Pop</span> <var>script execution context</var> from the <span>JavaScript execution context stack</span>.</p></li>
    <li><p><span>Clean up after running a callback</span>, `callback`.</p></li>
    <li><p>Return undefined.</li>
   </ol>
   <p>If an exception gets thrown by the callback <var>callback</var>, end these steps, and <span>report the exception</span>.</p>
```

fserb replied:
It seems a bit verbose, but I trust you if you say it makes more sense. Done.

fserb replied:
I'm not sure what you mean by "performing the callback", I cleaned this, see if it still makes sense for you.

cvan commented on source:
```diff
@@ -88980,13 +88978,35 @@ dictionary <dfn>PromiseRejectionEventInit</dfn> : <span>EventInit</span> {
+    </ol>
+    <p class="note">Similar to the
+    <span data-x="nested browsing context">nested browsing contexts</span> note above, a user
+    agent can consider updating that Dedicated Worker rendering infrequently or never.</p>
```
`a user agent can consider updating that Dedicated Worker rendering infrequently or never.` -> `a user agent can determine the rate of rendering in the <code>Dedicated Worker</code>.` or `a user agent can determine to degrade gracefully the rendering in the <code>Dedicated Worker</code>.`?

fserb replied:
done.

cvan commented on source:
```diff
@@ -88980,13 +88978,35 @@ dictionary <dfn>PromiseRejectionEventInit</dfn> : <span>EventInit</span> {
+       <code>DedicatedWorkerGlobalScope</code>, passing in <var>now</var> as the
+       timestamp.</p></li>
+
+       <li><p>Update the rendering of that Dedicated Worker to reflect the current state.</p>
```
should `Dedicated Worker` have a `<code>` or `<span data-x>`?

fserb replied:
I don't think we can point to anything. I did `code` it for `DedicatedWorkerGlobalScope` though.

cvan commented on source:
```diff
@@ -88980,13 +88978,35 @@ dictionary <dfn>PromiseRejectionEventInit</dfn> : <span>EventInit</span> {
+       <code>Performance</code> object's <code data-x="dom-Performance-now">now()</code> method.
+       <ref spec=HRT></p></li>
+
+       <li><p><span>run the animation frame callbacks</span> for that
```
can `run` be capitalised as `Run`? (I believe it'll still be linked correctly?)

fserb replied:
it does! Done.

fserb commented:
Done. PTAL.

fserb commented:
@annevk: WPT for RAF is in too. Are we good here?

zcorpan commented:
@fserb have you started writing tests for this, or is that up for grabs?

fserb commented:
I submitted this: https://github.com/web-platform-tests/wpt/blob/master/workers/WorkerGlobalScope_requestAnimationFrame.htm a couple weeks ago. 

We could still use some tests for RAF + OffscreenCanvas interaction that I haven't done. Or anything else that people may have in mind. But I haven't done anything further for this.

annevk commented on source:
```diff
@@ -93288,45 +93307,58 @@ loadMySprites().then(runDemo);</pre>
-   <li><p>Let <var>document</var> be this <code>Window</code> object's <span
-   data-x="concept-document-window">associated <code>Document</code></span>.</p></li>
+Window includes AnimationFrameProvider;
+DedicatedWorkerGlobalScope includes AnimationFrameProvider;
```
Since this isn't Bikeshed, you'll need to annotate the IDL manually as was done originally. Otherwise there's no linking.

fserb replied:
Done.

annevk commented:
Thanks to a blink-dev thread I realized there's another fundamental problem here. This only works for dedicated workers that belong to an agent cluster that also contains a window. If a shared or service worker creates a dedicated worker none of this should work.

I wonder if that means we need additional IDL machinery to hide the APIs there? At the very least we need to add more conditionals.

fserb commented:
Implementation bugs:

[Edge](https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/17934399/)
[Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1469282)
[Safari](https://bugs.webkit.org/show_bug.cgi?id=186760)
Chrome already implements it.

fserb commented:
Would it be possible for us to address the new worker situation in a separate PR? Or do you think we need to do this here?

annevk commented:
Ideally we do it here as it also affects API exposure and what this API needs to do in dedicated workers that are not equipped to handle this.

(As with software we try to keep the master branch somewhat clean and have each individual commit result in something that works.)

fserb commented:
Do you think we could limit this by saying that `requestAnimationFrame` is available on `DedicatedWorkers` that are created from a `window` object"` and should throw an exception when that's not the case?

This way we don't need new IDL mechanisms, although it's a bit ugly.

annevk commented:
@fserb so you don't want to support window -> dedicated worker -> dedicated worker? I think that should work as that's all the same agent cluster. I'd rather not introduce yet another boundary.

It seems somewhat okay to not have a new IDL boundary, but we should maybe ask more people for opinions since as far as I know it's a new thing (though maybe other APIs already did this?).

fserb commented:
I wouldn't mind that those be included, I was just suggesting that we could address this outside IDL changes.

That said, Do you feel that `agent cluster` is the right abstraction for that? Let me try without it. Let's see how it looks.

fserb commented:
What do you think about this solution?

annevk commented:
That doesn't define:

* when it's tagged "invalid"
* scope
* the exception to be thrown
* an exception for the second method

I'm pretty sure that an agent cluster that contains a similar-origin window agent is the boundary we want here, however we define it. What I'd like input from others on is whether exposing the API on dedicated workers outside that boundary is acceptable, even though we'll likely never make them work.

cc @domenic

fserb commented:
I've defined the exceptions on both methods.
For the "when/scope", you think this is not enough?
```
  <p>An <code>AnimationFrameProvider</code> implementation is marked as <var>invalid</var> if the object that includes it was created from a scope that doesn't implement a valid
  <code>AnimationFrameProvider</code>.
```

fserb commented:
Changed the DOMException. Also changed the "invalid" to "not supported" to mirror the exception. I'm not sure what you meant by "built upon primitives". Could you help me here?

annevk commented:
The current definition makes it sound like it's implementation defined when something is "not supported". Instead, it should be tied to the agent/agent cluster or traversing the worker's global's owner set.

fserb commented:
I thought it would be enough to make it "recursive" like that:

```
An AnimationFrameProvider implementation is considered not supported if the object that includes it was created from a scope that doesn't implement a supported AnimationFrameProvider. AnimationFrameProvider must be supported on Window.
```

I was trying to avoid adding the concept of agent cluster from Ecma. But it seems `worker's owner set` is a thing. So maybe I can try that too.

annevk commented:
@fserb agent cluster and similar-origin window agent and other terms I've been using are already a thing in HTML. The problem with your definition is that it doesn't define "scope".

fserb commented:
Does `execution context` works here instead?

annevk commented:
I don't see how. It doesn't seem to follow from this definition on what dedicated workers it functions?

fserb commented:
If the `DedicatedWorker` was created from a `Window` or from another `DedicatedWorker` that had a valid `requestAnimationFrame` then it will have. If a `DedicatedWorker` was created from a Service Worker, then it won't, because the execution context of the Service Worker doesn't have `requestAnimationFrame`.

annevk commented:
That doesn't work as the object that includes `AnimationFrameProvider` is the global of the dedicated worker and that's not created by the global of the `Window` object.

fserb commented:
okey.

I don't understand how to phrase this in terms of `owner set` because the `Document` is part of it, not the `Window`. Would it be enough to say that if the `owner set` contains a `Document` then the `DedicatedWork` must support `AnimationFrameProvider`?

annevk commented:
Yeah, but you need to make it recursive. If owner set only contains dedicated workers you want to look at their owners. I guess you can bail out if you have multiple owners too.

I think a more ideal solution would be that we introduce "current agent cluster" and we do a check based on that, maybe file a follow-up on that? Seems useful if this becomes more widespread.

fserb commented:
I gave it a try at the `owner set` language. PTAL.

annevk commented:
Thanks, I think that works, but it could use some editorial work that @domenic is hopefully willing to pick up. I'll be out for 3.5 weeks. I'm happy to pick this up when I'm back and try to finish it up if he doesn't get to it though. Note that I also wrote https://github.com/w3ctag/design-reviews/issues/288#issuecomment-398672236 to see if we could get some additional review on the overall implications here, but no feedback thus far.

wanderview commented:
Do workers attached to background windows fire rAF callbacks in the worker?

AIUI at least a few browsers do not fire rAF callbacks for background windows on the main thread.  Also, there is a desire to throttle background workers in general.

I see the spec has vague wording about "the user agent believes that it would benefit from having its rendering updated", but it was unclear to me if in practice this background throttling would apply to these workers.

domenic commented:
@fserb I did a pass through to fix up the various issues I found. Please look at the commit message on my latest push for details.

For most they were editorial, or the fix was straightforward. But the one that was really puzzling were the changes you made to "run the animation frame callbacks", which I didn't understand the reasoning for. I reverted them because they seemed likely to be wrong, but I'd love to understand what the thinking was there.

Please take a look at my changes and make sure they're aligned with what you had in mind :).

@wanderview I think currently the spec is trying to leave framerate, background throttling, etc. up to UAs, both for the window and worker cases. Are you asking about Chrome's planned implementation?

domenic commented:
Also, it would be good to have a small bit of sample code, showing how to use the worker's requestAnimationFrame, in the spec. I think that will be non-obvious to many readers, since this section never mentions any connection with OffscreenCanvas. (Which I assume is how this is all connected?)

fserb commented:
Added an example.

annevk commented:
The changeset includes lots of unrelated changes. Can you perhaps rebase against master and force push?

rniwa commented:
@smfr @grorg

fserb commented:
Yep. The WPT tests are in.

fserb commented:
@domenic, did I accidentally removed something that you changed? I'm very sorry if I did.

domenic commented:
Yeah, I can re-push it.

Can you link to the web platform tests?

fserb commented:
I'm sorry. :(

https://github.com/web-platform-tests/wpt/pull/11127

annevk commented on source:
```diff
@@ -89132,13 +89130,39 @@ dictionary <dfn>PromiseRejectionEventInit</dfn> : <span>EventInit</span> {
+    <ol>
+     <li>
+      <p>If this is a <code>DedicatedWorkerGlobalScope</code> and the user agent believes that
+      it would benefit from having its rendering updated at this time, then:</p>
```
This should also check the owner set to ensure it's an appropriate dedicated worker and not one that cannot have this functionality, right?

fserb replied:
Done.

annevk commented:
@fserb that still doesn't work. As I understand it all `DedicatedWorkerGlobalScope` objects will include this interface mixin. However, for only a subset of them will it actually function. The same check that stops the API from working needs to be made in the "update the rendering" step too. That should probably be a shared operation.

domenic commented:
Ugh, my commit at https://github.com/whatwg/html/pull/3677/commits/3a1946acce1d866e6b1326119af4a727c9edaf5d appears to have reverted a lot of stuff that made it into master in https://github.com/whatwg/html/commit/753e9fe231b903f88f6d92765a90d254b23347fc. Trying to fix now.

domenic commented:
OK, fixed that Git issue. Please do `git pull --rebase` before adding any more commits :).

I think what I'm hearing from @annevk's comment is that the check should also reference the "not supported" concept?

fserb commented:
I understood that. That's what I tried to do with `that implements a supported <code>AnimationFrameProvider</code>`. Unless you want me to write `not "not supported"`. But that sounds silly, no?

domenic commented:
I think the way to do that is to say

> If this WorkerGlobalScope is [supported] and the user agent believes...

where "supported" links to the "not supported" definition.

It might be better to actually define "supported" by inverting the conditions, though...

fserb commented:
Defined supported and linked it.

annevk commented on source:
```diff
@@ -89210,13 +89208,41 @@ dictionary <dfn>PromiseRejectionEventInit</dfn> : <span>EventInit</span> {
-   <a href="#workers">Web workers</a> section below.</p></li>
+   <li>
+    <p>If this is a <a href="#workers">worker</a> <span>event loop</span> (i.e., one running for
+    a <code>WorkerGlobalScope</code>):</p>
```
a WorkerGlobalScope object*

annevk replied:
(throughout)

domenic replied:
The existing text used this phrasing, and I find it simpler...

annevk replied:
We don't say "a <code>legend</code>" either and assume folks can infer " element". But I guess I'm fine with a follow-up.

annevk commented on source:
```diff
@@ -93816,67 +93842,135 @@ loadMySprites().then(runDemo);</code></pre>
+  unsigned long <span data-x="AnimationFrameProvider-requestAnimationFrame">requestAnimationFrame</span>(<span>FrameRequestCallback</span> callback);
+  void <span data-x="AnimationFrameProvider-cancelAnimationFrame">cancelAnimationFrame</span>(unsigned long handle);
+};
 
```
Please drop this newline. At least, I think the convention is that includes are directly adjacent to the mixin (or interface that includes them). Which begs the question, should these includes be listed below the Window and DedicatedWorkerGlobalScope interfaces instead? @domenic?

domenic replied:
I guess you're right about the newline. I think it's reasonable to have them here. The spec goes both ways, but this feels simpler.

annevk commented on source:
```diff
@@ -93816,67 +93842,135 @@ loadMySprites().then(runDemo);</code></pre>
-   <li><p>Increment <var>document</var>'s <span>animation frame callback identifier</span> by
-   one.</p></li>
+  <p>Each <code>AnimationFrameProvider</code> object also has a <dfn
+  data-x="concept-animationframeprovider-target-object">target object</dfn> that stores the
```
In data-x we preserve the capitalization usually, so concept-AnimationFrameProvider-target-object.

domenic replied:
It doesn't matter though; it's case-insensitive and gets converted to lowercase in the output. I'm not sure it's worth asking people to change.

annevk replied:
I think it's good to be consistent. It's the same kind of reason we wrap at a 100 columns.

annevk commented on source:
```diff
@@ -93816,67 +93842,135 @@ loadMySprites().then(runDemo);</code></pre>
-   callbacks</span>, associated with <var>document</var>'s <span>animation frame callback
-   identifier</span>'s current value.</p></li>
+  <dl>
+   <dt>If the <code>AnimationFrameProvider</code> is a <code>Window</code></dt>
```
It's not a Window, it's included by a Window object.

annevk replied:
(note, this also applies below)

domenic replied:
I disagree. The object in question is a Window object. Window as an interface includes the AnimationFrameProvider interface mixin, but the object is not included by a Window object.

Then it comes down to whether you think it's important to suffix everything with " object". I prefer not to, personally.

annevk replied:
I guess the problem is earlier on where it talks about an "AnimationFrameProvider object" (note that there object is included; are you saying it should be removed there?). I think that instead should talk about objects (implementing an interface that) including the AnimationFrameProvider interface mixin. Perhaps introducing _object_ as a variable for such objects and then here we can branch on that variable.

domenic replied:
My take is that "object" is an optional disambiguator. Here we stated it once, but don't have to repeat ourselves.

As for whether it's OK to talk about "X objects" at all for interface mixins, let's resolve that at https://github.com/heycam/webidl/issues/582, and leave this as-is for now.

annevk commented on source:
```diff
@@ -93816,67 +93842,135 @@ loadMySprites().then(runDemo);</code></pre>
-  <p>When the <dfn><code
-  data-x="dom-window-cancelAnimationFrame">cancelAnimationFrame()</code></dfn> method is called,
-  the user agent must run the following steps:</p>
+  <p>An <code>AnimationFrameProvider</code> <var>provider</var> is considered <dfn
```
I don't think this makes sense given that it's just an interface mixin. I guess "an object including the AnimationFrameProvider interface mixin _provider_ ..."?

domenic replied:
I think the phrasing here is an acceptable shorthand for your longhand.

Also, objects don't include mixins, so the longhand would actually have to be "an object _provider_ that implements an interface that includes the AnimationFrameProvider interface mixin".

annevk commented on source:
```diff
@@ -93816,67 +93842,135 @@ loadMySprites().then(runDemo);</code></pre>
+   <li><var>provider</var> is a <code>Window</code>.</li>
+
+   <li><var>provider</var>'s <span>owner set</span> <span data-x="list
+   contains">contain</span> a <code>Document</code> object.</li>
```
contains*

annevk commented on source:
```diff
@@ -93816,67 +93842,135 @@ loadMySprites().then(runDemo);</code></pre>
+   <li><var>provider</var>'s <span>owner set</span> <span data-x="list
+   contains">contain</span> a <code>Document</code> object.</li>
+
+   <li>Any of the <code>DedicatedWorkerGlobalScope</code> instances in <var>provider</var>'s
```
instances -> objects

domenic replied:
Why? Both instances and objects seem like reasonable shorthands for "objects that are instances of the X interface".

annevk replied:
I'm pretty sure we normally avoid "instance".

annevk commented on source:
```diff
@@ -93816,67 +93842,135 @@ loadMySprites().then(runDemo);</code></pre>
+
+  <p>The <dfn><code
+  data-x="AnimationFrameProvider-requestAnimationFrame">requestAnimationFrame(<var>callback</var>)</code></dfn>
+  method must run the following steps:</p>
```
method, when invoked, must*

domenic replied:
Why? We use the shorter phrasing 15 times in the spec, and shorter seems better.

annevk replied:
Is it? The steps are only to be run if the method is actually invoked. I guess I'm okay with letting this be though.

annevk commented on source:
```diff
@@ -93816,67 +93842,135 @@ loadMySprites().then(runDemo);</code></pre>
-  steps:</p>
+  <p>The <dfn><code
+  data-x="AnimationFrameProvider-cancelAnimationFrame">cancelAnimationFrame(<var>handle</var>)</code></dfn>
+  method must run the following steps:</p>
```
method, when invoked, must*

fserb commented:
updates?

domenic commented:
@annevk addressed most of your nits, with https://github.com/heycam/webidl/issues/582 being the remaining point of disagreement. I'd love to merge this; are you able to sign off?

annevk commented:
I pushed more nits.

I don't quite get why your revision failed to build as I cannot find a single instance of "list of animation frame callbacks" (including when I ignore the actual whitespace used) left. I guess we'll see if mine ends up building without error...

(I mistakenly added "needs tests" as I forgot about that conversion and a quick find didn't reveal WPT references, but that was due to GitHub eliding a whole bunch of comments.)

fserb commented:
Fixed the build problem. :)
